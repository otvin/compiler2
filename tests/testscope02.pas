program testscope02(output);
{tests combination of byref and global var scope}

var globalint:integer; globalreal:real; tmp:integer;

function byRef(var a:integer; var b:real):integer;
begin
  writeln('Before ');
  writeln(a);
  writeln(globalint);

  a := a + 2;
  globalint := globalint + 2;

  writeln('After ');
  writeln(a);
  writeln(globalint);

  writeln('Before Real');
  writeln(b);
  writeln(globalreal);

  b := b * 2;
  globalreal := globalreal * 2;

  writeln('After Real');
  writeln(b);
  writeln(globalreal);

  byRef := 0;
end;

function byVal(a:integer; b:real):integer;
begin
  writeln('Before ');
  writeln(a);
  writeln(globalint);

  a := a + 2;
  globalint := globalint + 2;

  writeln('After ');
  writeln(a);
  writeln(globalint);

  writeln('Before Real');
  writeln(b);
  writeln(globalreal);

  b := b * 2;
  globalreal := globalreal * 2;

  writeln('After Real');
  writeln(b);
  writeln(globalreal);

  byVal := 0;
end;

begin {main}
  globalint := 9;
  globalreal := 3.7;

  writeln('in main going in');
  writeln(globalint);
  writeln(globalreal);
  tmp := byval(globalint, globalreal);
  writeln('in main coming out');
  writeln(globalint);
  writeln(globalreal);

  writeln('in main going in2');
  writeln(globalint);
  writeln(globalreal);
  tmp := byref(globalint, globalreal);
  writeln('in main coming out2');
  writeln(globalint);
  writeln(globalreal);

  writeln('in main going in3');
  writeln(globalint);
  writeln(globalreal);
  tmp := byval(globalint, globalreal);
  writeln('in main coming out3');
  writeln(globalint);
  writeln(globalreal);
end.
