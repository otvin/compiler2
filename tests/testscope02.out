in main going in
9
3.700000000000
Before 
9
9
After 
11
11
Before Real
3.700000000000
3.700000000000
After Real
7.400000000000
7.400000000000
in main coming out
11
7.400000000000
in main going in2
11
7.400000000000
Before 
11
11
After 
15
15
Before Real
7.400000000000
7.400000000000
After Real
29.600000000000
29.600000000000
in main coming out2
15
29.600000000000
in main going in3
15
29.600000000000
Before 
15
15
After 
17
17
Before Real
29.600000000000
29.600000000000
After Real
59.200000000000
59.200000000000
in main coming out3
17
59.200000000000
